### ISSUE TITLE
Version 3.101.0 - `Pulumi refresh` => error: getting snapshot: constructing secrets manager of type "cloud": secrets (code=InvalidArgument)

### ISSUE THREAD

**Issue by m-yavuz** [pulumi/pulumi#15127]
### What happened?

On the latest v3.101.0 version of the CLI in our Azure DevOps pipeline, 
`Pulumi refresh`  and `Pulumi refresh` are not working

PS: everything was work on previous version 

### Example

When run `Pulumi refresh` or `Pulumi up`
I got this error 

```
error: getting snapshot: constructing secrets manager of type "cloud": secrets (code=InvalidArgument): POST https://someproject.vault.azure.net/keys/pulumi-secret-someproject-npr/decrypt
--------------------------------------------------------------------------------

RESPONSE 400: 400 Bad Request
ERROR CODE: BadParameter

--------------------------------------------------------------------------------
{
  "error": {
    "code": "BadParameter",
    "message": "The parameter is incorrect.\r\n"
  }
}
--------------------------------------------------------------------------------
```

### Output of `pulumi about`

**Cloud Provider : Azure**
**Pulumi version : v3.101.0**

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

**Comment by tgummerer** [comment]
This looks unfortunately related to https://github.com/pulumi/pulumi/pull/15011, which we're working on reverting in #15128.  We'll try to work out what happened exactly before trying that upgrade again.  Thanks for the report!

**Comment by tgummerer** [comment]
Closing this out, as we released Pulumi 3.101.1 which addresses this.  I opened https://github.com/pulumi/pulumi/issues/15138 to investigate what exactly happened and to attempt another upgrade once we figured out what caused this issue.

**Comment by justinvp** [comment]
Heads up that the next minor version of Pulumi (tentatively v3.104.0 to be released on 1/31/2024) will use the latest version of [Go Cloud Development Kit](https://gocloud.dev). We believe the error from this issue has been addressed with a regression test in place.

If you'd like to try the new CLI before it is officially released next week, you can install a dev build with the following commands:

## Linux / macOS

```sh
curl -sSL https://get.pulumi.com | sh -s -- --version dev
```

## Windows

Open a new command prompt windows (<kbd>WIN</kbd>+<kbd>R</kbd>: `cmd.exe`)

```bat
@"%SystemRoot%\System32\WindowsPowerShell\v1.0\powershell.exe" -NoProfile -InputFormat None -ExecutionPolicy Bypass -Command "[Net.ServicePointManager]::SecurityProtocol = [Net.SecurityProtocolType]::Tls12; iex ((New-Object System.Net.WebClient).DownloadString('https://get.pulumi.com/install.ps1')) -version dev" && SET "PATH=%PATH%;%USERPROFILE%\.pulumi\bin"
```

## GitHub Action

```yaml
- uses: pulumi/actions@v5
  with:
    pulumi-version: dev
```

If you try the dev build and do encounter a problem, please open a new issue at https://github.com/pulumi/pulumi/issues/new/choose

Thank you!
